node/pkg/reporter/node.go

```diff
@@ -0,0 +1,178 @@
+}
+
+type GlobalAggregate struct {
+	Name      string    `db:"name"`
+	Value     int64     `db:"value"`
+	Round     int64     `db:"round"`
+	Timestamp time.Time `db:"timestamp"`
+}
+
+func New(ctx context.Context, h host.Host, ps *pubsub.PubSub) (*Reporter, error) {
+	encryptedTopic, err := utils.EncryptText(TOPIC_STRING)
+	if err != nil {
+		return nil, err
+	}
+
+	topic, err := ps.Join(encryptedTopic)
+	if err != nil {
+		return nil, err
+	}
+
+	raft := raft.NewRaftNode(h, ps, topic, 100, LEADER_TIMEOUT)
+	txHelper, err := utils.NewTxHelper(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	contractAddress := os.Getenv("SUBMISSION_PROXY_CONTRACT")
+	if contractAddress == "" {
+		return nil, errors.New("SUBMISSION_PROXY_CONTRACT not set")
+	}
+
+	reporter := &Reporter{
+		Raft:            raft,
+		TxHelper:        txHelper,
+		contractAddress: contractAddress,
+		lastSubmissions: make(map[string]int64),
+	}
+	reporter.Raft.LeaderJob = reporter.leaderJob
+	reporter.Raft.HandleCustomMessage = reporter.handleCustomMessage
+
+	return reporter, nil
+}
+
+func (r *Reporter) Run(ctx context.Context) {
+	r.Raft.Run(ctx)
+}
+
+func (r *Reporter) leaderJob() error {
+	ctx := context.Background()
+	failureTimout := 50 * time.Millisecond
+	for i := 0; i < MAX_RETRY; i++ {
+		failureTimout += failureTimout + time.Duration(rand.Intn(100))*time.Millisecond
+		if failureTimout > 500*time.Millisecond {
+			failureTimout = 500 * time.Millisecond
+		}
+
+		aggregates, err := r.getLatestGlobalAggregates(ctx)
+		if err != nil {
+			log.Error().Err(err).Msg("GetLatestGlobalAggregates")
+			time.Sleep(failureTimout)
+			continue
+		}
+
+		validAggregates := r.filterInvalidAggregates(aggregates)
+		if len(validAggregates) == 0 {
+			log.Error().Msg("no valid aggregates to report")
+			time.Sleep(failureTimout)
+			continue
+		}
+
+		err = r.report(ctx, validAggregates)
+		if err != nil {
+			log.Error().Err(err).Msg("Report")
+			time.Sleep(failureTimout)
+			continue
+		}
+
+		for _, agg := range validAggregates {
+			r.lastSubmissions[agg.Name] = agg.Round
+		}
+		return nil
+	}
+	r.resignLeader()
+	return errors.New("failed to report")
+}
+
+func (r *Reporter) resignLeader() {
+	r.Raft.StopHeartbeatTicker()
+	r.Raft.UpdateRole(raft.Follower)
+}
+
+func (r *Reporter) handleCustomMessage(msg raft.Message) error {
+	// TODO: implement message handling related to validation
+	return errors.New("unknown message type")
+}
+
+func (r *Reporter) getLatestGlobalAggregates(ctx context.Context) ([]GlobalAggregate, error) {
+	return db.QueryRows[GlobalAggregate](ctx, GET_LATEST_GLOBAL_AGGREGATES_QUERY, nil)
+}
+
+func (r *Reporter) report(ctx context.Context, aggregates []GlobalAggregate) error {
+	pairs, values := r.makeContractArgs(aggregates)
+	rawTx, err := r.TxHelper.MakeDirectTx(ctx, r.contractAddress, FUNCTION_STRING, pairs, values)
+	if err != nil {
+		log.Error().Err(err).Msg("MakeDirectTx")
+		return err
+	}
+	return r.TxHelper.SubmitRawTx(ctx, rawTx)
```

Comment:
receipt is logged from `SubmitRawTx` function so adding more logging is unnecessary, in case of other failures, it's handled from TxHelper package, it'll return error on failure